lib/manager/components/validation/ServicePerModeChart.js

```diff
@@ -15,46 +14,47 @@ type Props = {
 }
 
 const maxDaysToShow = 500
+const orderedPossibleModes = ['other', 'metro', 'rail', 'tram', 'bus']
 
 export default class ServicePerModeChart extends Component<Props> {
-  _getData = (validationResult: ValidationResult) => {
+  _getData = (validationResult: ValidationResult): {
+    data: Array<{
+      bus: number,
+      date: moment,
+      metro: number,
+      other: number,
+      rail: number,
+      total: number,
+      tram: number
+    }>,
+    modes: Array<string>
+  } => {
     const {dailyBusSeconds, dailyTramSeconds, dailyMetroSeconds, dailyRailSeconds, dailyTotalSeconds, firstCalendarDate} = validationResult
     const firstDate = moment(firstCalendarDate)
-    const hasBus = dailyBusSeconds.filter(s => s > 0).length > 0
-    const hasTram = dailyTramSeconds.filter(s => s > 0).length > 0
-    const hasMetro = dailyMetroSeconds.filter(s => s > 0).length > 0
-    const hasRail = dailyRailSeconds.filter(s => s > 0).length > 0
-    let hasOther = false
     const data = []
+    const modes = {}
     const limit = Math.min(dailyTotalSeconds.length, maxDaysToShow)
     for (let i = 0; i < limit; i++) {
       const column = {}
       column.date = firstDate.clone().add(i, 'days')
-      let other = dailyTotalSeconds[i]
       column.total = this._secondsToHours(dailyTotalSeconds[i]) // total
-      if (hasBus) {
-        column.bus = this._secondsToHours(dailyBusSeconds[i]) // bus
-        other -= dailyBusSeconds[i]
-      }
-      if (hasMetro) {
-        column.metro = this._secondsToHours(dailyMetroSeconds[i]) // metro
-        other -= dailyMetroSeconds[i]
-      }
-      if (hasRail) {
-        column.rail = this._secondsToHours(dailyRailSeconds[i]) // rail
-        other -= dailyRailSeconds[i]
-      }
-      if (hasTram) {
-        column.tram = this._secondsToHours(dailyTramSeconds[i]) // tram
-        other -= dailyTramSeconds[i]
-      }
-      if (other || hasOther) {
-        column.other = this._secondsToHours(other)
-        hasOther = true
-      }
+      column.bus = this._secondsToHours(dailyBusSeconds[i]) // bus
+      column.metro = this._secondsToHours(dailyMetroSeconds[i]) // metro
+      column.rail = this._secondsToHours(dailyRailSeconds[i]) // rail
+      column.tram = this._secondsToHours(dailyTramSeconds[i]) // tram
+      column.other = column.total -
+        column.bus -
+        column.metro -
+        column.rail -
+        column.tram
+      orderedPossibleModes.forEach(mode => {
+        if (column[mode] > 0) {
+          modes[mode] = true
+        }
+      })
```

Comment:
Why not just do this check for modes with value > 0 in the filter below after the `this._getData` call? The way it currently is feels a little obfuscated. Alternatively, if you decide to keep it this way, the `modes` key could use a more descriptive name e.g., `modesWithData` or something.

Reply:
I ended up doing the check in the filter as you recommended.